File: apps/go-api/internal/analysis/weapontier/weapontier_test.go
package weapontier

import (
	"testing"
)

// Familles d'arme du témoin, écrites comme le film les écrit.
const (
	ar     = "0x48C19D2D" // MA40 — arme de départ
	pistol = "0xF408190F" // Sidekick — arme de départ
	sniper = "0x9D6AAED2" // S7 — socle de puissance
	hydra  = "0xB619D84A" // Hydra — râtelier, malgré son rôle `power`
	bonus  = "powerup_overshield"
)

// temoin — quatre socles : un râtelier, un socle de puissance, un bonus, et un socle
// qu'aucun emplacement ne confirme.
func temoin() ([]Pad, []Spot) {
	pads := []Pad{
		{Weapon: hydra},  // 0 — râtelier
		{Weapon: sniper}, // 1 — puissance
		{Weapon: bonus},  // 2 — bonus
		{Weapon: ar},     // 3 — NON confirmé
	}
	cross := []Spot{
		{Pad: 0, Family: "rack"},
		{Pad: 1, Family: "power"},
		{Pad: 2, Family: "powerup"},
	}
	return pads, cross
}

// vies fabrique un canal `loadouts` : n vies portant `w`, chacune sur son propre slot, avec
// une SECONDE émission par vie (le canal ré-émet après un changement d'arme) pour que les
// tests prouvent que seule la première compte.
func vies(depart int, w []string, apres []string) []Spawn {
	var out []Spawn
	for i := 0; i < depart; i++ {
		slot := uint32(512 + i)
		out = append(out, Spawn{Slot: slot, Weapons: w})
		if apres != nil {
			out = append(out, Spawn{Slot: slot, Weapons: apres})
		}
	}
	return out
}

func TestTierOf_LesQuatreNiveaux(t *testing.T) {
	pads, cross := temoin()
	m := NewMatch(pads, cross, vies(20, []string{ar, pistol}, []string{sniper, pistol}), false)

	cas := []struct {
		nom  string
		pad  int
		arme string
		want Tier
	}{
		{"râtelier", 0, hydra, TierGround},
		{"socle de puissance", 1, sniper, TierPower},
		{"socle de bonus", 2, bonus, TierPowerup},
		{"socle non confirmé, arme de base", 3, ar, TierBase},
	}
	for _, c := range cas {
		if got := m.TierOf(c.pad, c.arme); got != c.want {
			t.Errorf("%s : niveau = %q, attendu %q", c.nom, got, c.want)
		}
	}
	if m.Lives() != 20 {
		t.Errorf("vies lues = %d, attendu 20", m.Lives())
	}
}

// TestTierOf_NonClasseQuandRienNeConfirme — D4 : un socle sans emplacement confirmé reste
// non classé et visible ; il ne tombe JAMAIS dans « terrain » par défaut.
func TestTierOf_NonClasseQuandRienNeConfirme(t *testing.T) {
	pads, _ := temoin()
	m := NewMatch(pads, nil, vies(20, []string{ar, pistol}, nil), false)
	if got := m.TierOf(1, sniper); got != TierUnclassified {
		t.Errorf("sans référence de carte : niveau = %q, attendu %q", got, TierUnclassified)
	}
	// Et un index de socle hors bornes ne pioche pas le voisin.
	if got := m.TierOf(99, sniper); got != TierUnclassified {
		t.Errorf("index hors bornes : niveau = %q, attendu %q", got, TierUnclassified)
	}
}

// TestTierOf_BasePrimeSurLEmplacement — l'ordre du plan : base > terrain > puissance.
func TestTierOf_BasePrimeSurLEmplacement(t *testing.T) {
	pads := []Pad{{Weapon: ar}}
	cross := []Spot{{Pad: 0, Family: "rack"}}
	m := NewMatch(pads, cross, vies(20, []string{ar, pistol}, nil), false)
	if got := m.TierOf(0, ar); got != TierBase {
		t.Errorf("arme de départ sur râtelier : niveau = %q, attendu %q", got, TierBase)
	}
}

// TestBaseWeapons_SeulePremiereEmission — le piège mesuré à l'étape 0 : le canal ré-émet.
func TestBaseWeapons_SeulePremiereEmission(t *testing.T) {
	pads := []Pad{{Weapon: sniper}}
	cross := []Spot{{Pad: 0, Family: "power"}}
	// Les vingt vies démarrent AR+Sidekick puis ramassent toutes le sniper : s'il comptait,
	// le sniper serait « base » et le niveau « puissance » disparaîtrait du match.
	m := NewMatch(pads, cross, vies(20, []string{ar, pistol}, []string{sniper, pistol}), false)
	if got := m.TierOf(0, sniper); got != TierPower {
		t.Errorf("arme ramassée en cours de vie : niveau = %q, attendu %q", got, TierPower)
	}
}

// TestBaseWeapons_SeuilDeQueue — la queue de 5,6 % mesurée à l'étape 0 ne promeut rien.
func TestBaseWeapons_SeuilDeQueue(t *testing.T) {
	pads := []Pad{{Weapon: sniper}}
	cross := []Spot{{Pad: 0, Family: "power"}}
	l := vies(40, []string{ar, pistol}, nil)
	// Une seule vie sur 41 démarre sniper en main : 2,4 %, sous BaseShareMin.
	l = append(l, Spawn{Slot: 999, Weapons: []string{sniper, pistol}})
	m := NewMatch(pads, cross, l, false)
	if got := m.TierOf(0, sniper); got != TierPower {
		t.Errorf("arme de la queue : niveau = %q, attendu %q", got, TierPower)
	}
	// Douze vies sur 52 (23 %) suffisent en revanche.
	for i := 0; i < 11; i++ {
		l = append(l, Spawn{Slot: uint32(900 + i), Weapons: []string{sniper, pistol}})
	}
	if got := NewMatch(pads, cross, l, false).TierOf(0, sniper); got != TierBase {
		t.Errorf("arme de départ franche : niveau = %q, attendu %q", got, TierBase)
	}
}

// TestTierOf_DepartsAleatoires — Fiesta : aucun niveau « base », les deux autres restent.
func TestTierOf_DepartsAleatoires(t *testing.T) {
	pads, cross := temoin()
	// En Fiesta l'AR est bien distribué au départ, mais le niveau ne doit pas exister.
	m := NewMatch(pads, cross, vies(20, []string{ar, pistol}, nil), true)
	if !m.RandomStarts() {
		t.Fatal("RandomStarts devrait être vrai")
	}
	if got := m.TierOf(3, ar); got != TierUnclassified {
		t.Errorf("départs aléatoires, socle non confirmé : niveau = %q, attendu %q", got, TierUnclassified)
	}
	if got := m.TierOf(0, hydra); got != TierGround {
		t.Errorf("départs aléatoires, râtelier : niveau = %q, attendu %q", got, TierGround)
	}
	if got := m.TierOf(1, sniper); got != TierPower {
		t.Errorf("départs aléatoires, puissance : niveau = %q, attendu %q", got, TierPower)
	}
	// Les vies restent comptées : l'écran doit pouvoir dire sur quoi il s'appuie.
	if m.Lives() != 20 {
		t.Errorf("vies lues = %d, attendu 20", m.Lives())
	}
}

// TestMatchZero_Utilisable — le zéro classe tout en non classé sans paniquer.
func TestMatchZero_Utilisable(t *testing.T) {
	var m Match
	if got := m.TierOf(0, ar); got != TierUnclassified {
		t.Errorf("Match zéro : niveau = %q, attendu %q", got, TierUnclassified)
	}
	if m.Lives() != 0 || m.RandomStarts() {
		t.Errorf("Match zéro : vies=%d aléatoire=%v", m.Lives(), m.RandomStarts())
	}
}

func TestCrossCheck_UnSeulSens(t *testing.T) {
	pads, cross := temoin()
	roles := map[string]string{hydra: "power", sniper: "sniper", ar: "automatic", pistol: "sidearm"}
	roleOf := func(w string) string { return roles[w] }

	m := NewMatch(pads, cross, nil, false)
	c := m.RunCrossCheck(pads, roleOf)
	// L'Hydra (rôle `power`) sur râtelier NE COMPTE PAS : c'est le cas nominal à 10,5 %.
	if c.LightOnPower != 0 {
		t.Errorf("aucun rôle léger sur puissance attendu, obtenu %d (%v)", c.LightOnPower, c.Weapons)
	}
	if c.Pads != 4 {
		t.Errorf("socles examinés = %d, attendu 4", c.Pads)
	}
	if c.Alert() {
		t.Error("alerte levée sur un match sain")
	}

	// Un AR (rôle `automatic`) sur le socle de puissance : c'est le sens suspect.
	pads[1] = Pad{Weapon: ar}
	c = m.RunCrossCheck(pads, roleOf)
	if c.LightOnPower != 1 || c.Weapons[ar] != 1 {
		t.Fatalf("inversion non comptée : %+v", c)
	}
	// 1 sur 4 = 25 %, très au-dessus des 2 %.
	if !c.Alert() {
		t.Error("alerte non levée à 25 % de socles inversés")
	}
}

// TestCrossCheck_ShotgunNestPasLeger — le fusil à pompe EST une arme de socle de puissance.
func TestCrossCheck_ShotgunNestPasLeger(t *testing.T) {
	pads := []Pad{{Weapon: "0xSHOT"}}
	cross := []Spot{{Pad: 0, Family: "power"}}
	m := NewMatch(pads, cross, nil, false)
	c := m.RunCrossCheck(pads, func(string) string { return "shotgun" })
	if c.LightOnPower != 0 || c.Alert() {
		t.Errorf("le fusil à pompe ne doit pas alerter : %+v", c)
	}
}

File: apps/web/src/features/match-replay/model/weaponTier.test.ts
/**
 * Tests — weaponTier (base / terrain / puissance / non classé).
 *
 * CE QU'ILS PROTÈGENT, dans l'ordre des pièges du domaine :
 *   - le niveau vient de la CARTE, jamais du nom ni du rôle de l'arme ;
 *   - un socle qu'aucun emplacement ne confirme reste NON CLASSÉ, jamais rangé ailleurs ;
 *   - « base » se lit sur la PREMIÈRE émission de chaque slot — le canal ré-émet en cours de
 *     vie, et tout prendre ferait d'une arme de puissance une arme de base ;
 *   - la queue d'équipements de départ (5,6 % mesurés) ne promeut rien : seuil `BASE_SHARE_MIN` ;
 *   - une émission qui SUIT une prise d'arme de la même vie n'est pas un départ (la grille
 *     d'images-clés est globale, pas alignée sur le spawn — cas `b1ad85eb` du 2026-09-21) ;
 *   - un mode à départs ALÉATOIRES ne publie aucun niveau « base », les deux autres restent.
 *
 * Ce fichier est le jumeau de `internal/analysis/weapontier/weapontier_test.go` : mêmes cas,
 * mêmes valeurs. Une divergence entre les deux est un bug, pas une variante.
 */
import { describe, expect, it } from 'vitest'

import type { ReplayDocument } from '@/lib/api/types'

import { buildPadTierMatch, padTierOf } from './weaponTier'
import { testReplayDoc } from '../test/testDoc'

const AR = '0x48C19D2D'
const PISTOLET = '0xF408190F'
const SNIPER = '0x9D6AAED2'
const HYDRA = '0xB619D84A'
const BONUS = 'powerup_overshield'

function socle(weapon: string) {
  return { weapon, x: 0, y: 0, spawns: [], presence: [] }
}

/** `n` vies partant avec `depart`, chacune sur son slot, plus une ré-émission si `apres`. */
function vies(n: number, depart: string[], apres?: string[]) {
  const out: { t: number; slot: number; w: string[] }[] = []
  for (let i = 0; i < n; i += 1) {
    out.push({ t: 74, slot: 512 + i, w: depart })
    if (apres) out.push({ t: 600, slot: 512 + i, w: apres })
  }
  return out
}

/**
 * LE TÉMOIN. Quatre socles : un râtelier, un socle de puissance, un socle de bonus, et un
 * quatrième qu'aucun emplacement ne confirme.
 */
function temoin(over: Partial<ReplayDocument> = {}) {
  return testReplayDoc({
    frameCount: 1000,
    frameIntervalMs: 100,
    weaponPads: [socle(HYDRA), socle(SNIPER), socle(BONUS), socle(AR)],
    mapWeaponPads: {
      catalogN: 9,
      pads: [
        { x: 0, y: 0, pad: 0, family: 'rack' },
        { x: 1, y: 0, pad: 1, family: 'power' },
        { x: 2, y: 0, pad: 2, family: 'powerup' },
      ],
    },
    loadouts: vies(20, [AR, PISTOLET], [SNIPER, PISTOLET]),
    ...over,
  } as Partial<ReplayDocument>)
}

/** Le même témoin, mais sur un mode à DÉPARTS ALÉATOIRES — le serveur le dit. */
function temoinAleatoire(over: Partial<ReplayDocument> = {}) {
  return temoin({ weaponTiers: { randomStarts: true }, ...over } as Partial<ReplayDocument>)
}

describe('padTierOf — les quatre niveaux', () => {
  it('lit la nature de l’emplacement sur la CARTE, et l’arme de base sur le film', () => {
    const m = buildPadTierMatch(temoin())
    expect(padTierOf(m, 0, HYDRA)).toBe('ground')
    expect(padTierOf(m, 1, SNIPER)).toBe('power')
    expect(padTierOf(m, 2, BONUS)).toBe('powerup')
    expect(padTierOf(m, 3, AR)).toBe('base')
    expect(m.lives).toBe(20)
    expect(m.tiersMeasured).toBe(true)
  })

  it('ne juge JAMAIS au rôle de l’arme : l’Hydra sur râtelier reste de terrain', () => {
    // L'Hydra a un rôle `power` au registre canonique, et 17 socles du parc la posent sur un
    // râtelier. Si le niveau se lisait sur l'arme, cette ligne rendrait 'power'.
    const m = buildPadTierMatch(temoin())
    expect(padTierOf(m, 0, HYDRA)).toBe('ground')
  })

  it('laisse NON CLASSÉ ce qu’aucun emplacement ne confirme, et le dit', () => {
    const m = buildPadTierMatch(temoin({ mapWeaponPads: undefined } as Partial<ReplayDocument>))
    expect(m.tiersMeasured).toBe(false)
    expect(padTierOf(m, 1, SNIPER)).toBe('unclassified')
    // Un index hors bornes ne pioche pas le voisin.
    expect(padTierOf(m, 99, SNIPER)).toBe('unclassified')
  })

  it('fait primer « base » sur l’emplacement (base > terrain > puissance)', () => {
    const m = buildPadTierMatch(
      temoin({
        weaponPads: [socle(AR)],
        mapWeaponPads: { catalogN: 1, pads: [{ x: 0, y: 0, pad: 0, family: 'rack' }] },
        loadouts: vies(20, [AR, PISTOLET]),
      } as Partial<ReplayDocument>),
    )
    expect(padTierOf(m, 0, AR)).toBe('base')
  })
})

describe('les armes de départ', () => {
  it('ne lit que la PREMIÈRE émission de chaque slot', () => {
    // Les vingt vies démarrent AR + Sidekick puis ramassent toutes le sniper. Si tout le canal
    // comptait, le sniper deviendrait « base » et le niveau « puissance » disparaîtrait.
    const m = buildPadTierMatch(temoin())
    expect(padTierOf(m, 1, SNIPER)).toBe('power')
  })

  it('ne promeut pas la queue : une vie sur 41 reste sous le seuil', () => {
    const doc = temoin({
      loadouts: [...vies(40, [AR, PISTOLET]), { t: 74, slot: 999, w: [SNIPER, PISTOLET] }],
    } as Partial<ReplayDocument>)
    expect(padTierOf(buildPadTierMatch(doc), 1, SNIPER)).toBe('power')
  })

  it('retient en revanche une arme de départ franche (12 vies sur 52)', () => {
    const enPlus = Array.from({ length: 12 }, (_, i) => ({
      t: 74,
      slot: 900 + i,
      w: [SNIPER, PISTOLET],
    }))
    const doc = temoin({
      loadouts: [...vies(40, [AR, PISTOLET]), ...enPlus],
    } as Partial<ReplayDocument>)
    expect(padTierOf(buildPadTierMatch(doc), 1, SNIPER)).toBe('base')
  })

  /**
   * LE CAS `b1ad85eb` (constat utilisateur du 2026-09-21), réduit à sa mécanique.
   *
   * Loadout de départ classique et ÉGAL pour tous, et pourtant l'Empaleur y était classé
   * « arme de base ». Cause mesurée : le canal `loadouts` est publié sur une grille
   * d'images-clés GLOBALE (t = 12, 212, 412 … toutes les 200 frames = 20 s), pas au spawn ;
   * cinq vies sur 73 avaient donc leur première émission APRÈS avoir ramassé un Empaleur —
   * 6,85 % des vies, au-dessus des 5 % du seuil.
   */
  it('écarte une émission POSTÉRIEURE à une prise d’arme de la même vie', () => {
    const prises = Array.from({ length: 5 }, (_, i) => ({
      t: 60,
      slot: 900 + i,
      w: '9d6aaed2',
      kind: 'weapon',
      class: 0,
    }))
    const emissions = Array.from({ length: 5 }, (_, i) => ({
      t: 200,
      slot: 900 + i,
      w: [SNIPER, PISTOLET],
    }))
    const pistes = Array.from({ length: 5 }, (_, i) => ({
      slot: 900 + i,
      team: 0,
      points: [{ t: 0, x: 0, y: 0 }],
    }))
    const commun = {
      loadouts: [...vies(40, [AR, PISTOLET]), ...emissions],
      tracks: pistes,
    } as Partial<ReplayDocument>
    // Témoin de la cause : sans les prises, les 5 vies pèsent 11,1 % et promeuvent le sniper.
    expect(padTierOf(buildPadTierMatch(temoin(commun)), 1, SNIPER)).toBe('base')
    const corrige = buildPadTierMatch(
      temoin({ ...commun, pickups: prises } as Partial<ReplayDocument>),
    )
    expect(padTierOf(corrige, 1, SNIPER)).toBe('power')
    // Les vies écartées quittent AUSSI le dénominateur : le silence n'est pas un zéro.
    expect(corrige.lives).toBe(40)
  })

  it('ne mord que dans un sens : une prise POSTÉRIEURE à l’émission ne disqualifie rien', () => {
    const prises = Array.from({ length: 20 }, (_, i) => ({
      t: 300,
      slot: 512 + i,
      w: '9d6aaed2',
      kind: 'weapon',
      class: 0,
    }))
    const m = buildPadTierMatch(
      temoin({
        loadouts: vies(20, [AR, PISTOLET]),
        pickups: prises,
        tracks: Array.from({ length: 20 }, (_, i) => ({
          slot: 512 + i,
          team: 0,
          points: [{ t: 0, x: 0, y: 0 }],
        })),
      } as Partial<ReplayDocument>),
    )
    expect(m.lives).toBe(20)
    expect(padTierOf(m, 3, AR)).toBe('base')
  })

  it('ne prend pas la DOTATION DE RÉAPPARITION pour une prise', () => {
    // `pickups` date la remise en main des armes de départ au début de la vie : huit prises à
    // t = 0 pour les huit premières vies de b1ad85eb. Sans le filtre du début de vie, les vies
    // les plus propres seraient les premières écartées.
    const m = buildPadTierMatch(
      temoin({
        loadouts: vies(20, [AR, PISTOLET]),
        pickups: Array.from({ length: 20 }, (_, i) => ({
          t: 0,
          slot: 512 + i,
          w: '48c19d2d',
          kind: 'weapon',
          class: 0,
        })),
        tracks: Array.from({ length: 20 }, (_, i) => ({
          slot: 512 + i,
          team: 0,
          points: [{ t: 0, x: 0, y: 0 }],
        })),
      } as Partial<ReplayDocument>),
    )
    expect(m.lives).toBe(20)
    expect(padTierOf(m, 3, AR)).toBe('base')
  })
})

describe('les modes à départs aléatoires', () => {
  it('lit le caractère aléatoire SUR LA RÉPONSE, jamais sur une liste locale', () => {
    // La règle vit dans le TOML du titre et gouverne aussi l'écriture en base : une copie ici
    // a existé une semaine et a divergé (revue du 2026-09-14).
    expect(buildPadTierMatch(temoinAleatoire()).randomStarts).toBe(true)
    expect(buildPadTierMatch(temoin()).randomStarts).toBe(false)
    // Champ ABSENT = le serveur ne sait pas : on ne publie pas la note, on ne l'invente pas.
    const sansChamp = temoin({ weaponTiers: undefined } as Partial<ReplayDocument>)
    expect(buildPadTierMatch(sansChamp).randomStarts).toBe(false)
  })

  it('ne publie AUCUN niveau « base », et garde les deux autres', () => {
    const m = buildPadTierMatch(temoinAleatoire())
    expect(m.randomStarts).toBe(true)
    // L'AR est bien distribué au départ en Fiesta, mais le niveau n'existe pas.
    expect(padTierOf(m, 3, AR)).toBe('unclassified')
    expect(padTierOf(m, 0, HYDRA)).toBe('ground')
    expect(padTierOf(m, 1, SNIPER)).toBe('power')
    // Les vies restent comptées : l'écran doit pouvoir dire sur quoi il s'appuie.
    expect(m.lives).toBe(20)
  })
})
